init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
@ngimel: Thanks for the comments!  
Yes I meant different thing: always reassign layer's extraBuffer from stream-shared one : self.extraBuffer = cudnn.getSharedWorkspace() - I am restoring per-stream shared buffers. 
As for the guesswork, I plan to eliminate it mostly by adding functionality into cutorch (memory pool so the workspace may be quickly released/acquired, a method that returns (a better guess about) largest available block and/or THCGetTempBuf(atLeastBytes, upToBytes).